File: src/main/java/me/itzisonn_/meazy/instruction/number/LogicalOperationInstruction.java
package me.itzisonn_.meazy.instruction.number;

import lombok.AllArgsConstructor;
import me.itzisonn_.meazy.instruction.Instruction;
import me.itzisonn_.meazy.instruction.BytecodeBuilders;
import org.jspecify.annotations.NullMarked;

import java.lang.classfile.CodeBuilder;

@NullMarked
@AllArgsConstructor
public final class LogicalOperationInstruction implements Instruction {
    private final LogicalOperation operation;

    @Override
    public void emit(BytecodeBuilders bytecodeBuilders) {
        CodeBuilder codeBuilder = bytecodeBuilders.getCodeBuilder();
        if (codeBuilder == null) throw new RuntimeException("Code builder is null");

        switch (operation) {
            case OR -> codeBuilder.ior();
            case AND -> codeBuilder.iand();
        }
    }

    public enum LogicalOperation {
        OR,
        AND
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/instruction/number/LogicalOperationInstruction.kt
package me.itzisonn_.meazy.instruction.number

import me.itzisonn_.meazy.instruction.BytecodeBuilders
import me.itzisonn_.meazy.instruction.Instruction

class LogicalOperationInstruction(private val operation: LogicalOperation) : Instruction {
    override fun emit(bytecodeBuilders: BytecodeBuilders) {
        val codeBuilder = bytecodeBuilders.codeBuilder ?: error("Code builder is null")

        when (operation) {
            LogicalOperation.OR -> codeBuilder.ior()
            LogicalOperation.AND -> codeBuilder.iand()
        }
    }

    enum class LogicalOperation {
        OR,
        AND
    }
}
